common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RetryingBlockTransferorSuite.java

```diff
@@ -365,13 +365,32 @@ public void testIOExceptionFailsConnectionEvenWithSaslException()
             ImmutableMap.of("b0", block0)
     );
     configMap.put("spark.shuffle.sasl.enableRetries", "true");
-    performInteractions(interactions, listener);
+    performInteractions(interactions, listener, false);
     verify(listener, timeout(5000)).onBlockTransferFailure("b0", saslExceptionFinal);
     verify(listener, atLeastOnce()).getTransferType();
     verifyNoMoreInteractions(listener);
     assert(_retryingBlockTransferor.getRetryCount() == MAX_RETRIES);
   }
 
+  @Test
+  public void testRetryInitiationFailure() throws IOException, InterruptedException {
+    BlockFetchingListener listener = mock(BlockFetchingListener.class);
+
+    List<? extends Map<String, Object>> interactions = Arrays.asList(
+        // IOException will initiate a retry, but the initiation will fail
+        ImmutableMap.<String, Object>builder()
+            .put("b0", new IOException("Connection failed or something"))
+            .put("b1", block1)
+            .build()
+    );
+
+    performInteractions(interactions, listener, true);
```

Comment:
Using the proposed changes below, this will become:

```suggestion
    configureInteractions(interactions, listener);
    _retryingBlockTransferor = spy(_retryingBlockTransferor);
    // Throw an OOM when initiating retries.
    doThrow(OutOfMemoryError.class).when(_retryingBlockTransferor).initiateRetry(any());
    // Override listener, so that it delegates to the spied instance and not the original class.
    _retryingBlockTransferor.setCurrentListener(_retryingBlockTransferor.new RetryingBlockTransferListener());
    _retryingBlockTransferor.start();
```